File: CS/Demo.Module/BusinessObjects/Address.cs
using System;
using DevExpress.Xpo;
using System.ComponentModel;
using System.Collections.Generic;
using DevExpress.Persistent.Base;
using GenerateUserFriendlyId.Module.BusinessObjects;

namespace Demo.Module.BusinessObjects {
    [DefaultClassOptions]
    [DefaultProperty("FullAddress")]
    [ImageName("BO_Address")]
    public class Address : UserFriendlyIdPersistentObject {
        string zipCode;
        string country;
        string province;
        string city;
        string address1;
        string address2;
        [PersistentAlias("Concat('A',PadLeft(ToStr(SequentialNumber),6,'0'))")]
        public string AddressId {
            get {
                return Convert.ToString(EvaluateAlias("AddressId"));
            }
        }
        public string Province {
            get {
                return province;
            }
            set {
                SetPropertyValue("Province", ref province, value);
            }
        }
        public string ZipCode {
            get {
                return zipCode;
            }
            set {
                SetPropertyValue("ZipCode", ref zipCode, value);
            }
        }
        public string Country {
            get {
                return country;
            }
            set {
                SetPropertyValue("Country", ref country, value);
            }
        }
        public string City {
            get {
                return city;
            }
            set {
                SetPropertyValue("City", ref city, value);
            }
        }
        public string Address1 {
            get {
                return address1;
            }
            set {
                SetPropertyValue("Address1", ref address1, value);
            }
        }
        public string Address2 {
            get {
                return address2;
            }
            set {
                SetPropertyValue("Address2", ref address2, value);
            }
        }
        [Association]
        public XPCollection<Contact> Persons {
            get {
                return GetCollection<Contact>("Persons");
            }
        }
        [PersistentAlias("concat(Country, Province, City, ZipCode)")]
        public string FullAddress {
            get { return ObjectFormatter.Format("{Country}; {Province}; {City}; {ZipCode}", this, EmptyEntriesMode.RemoveDelimiterWhenEntryIsEmpty); }
        }
        public Address(Session session) : base(session) { }
    }
}

File: VB/Demo.Module/BusinessObjects/Address.vb
Imports System
Imports DevExpress.Xpo
Imports System.ComponentModel
Imports DevExpress.Persistent.Base
Imports GenerateUserFriendlyId.Module.BusinessObjects

Namespace Demo.Module.BusinessObjects

    <DefaultClassOptions>
    <DefaultProperty("FullAddress")>
    <ImageName("BO_Address")>
    Public Class Address
        Inherits UserFriendlyIdPersistentObject

        Private zipCodeField As String

        Private countryField As String

        Private provinceField As String

        Private cityField As String

        Private address1Field As String

        Private address2Field As String

        <PersistentAlias("Concat('A',PadLeft(ToStr(SequentialNumber),6,'0'))")>
        Public ReadOnly Property AddressId As String
            Get
                Return Convert.ToString(EvaluateAlias("AddressId"))
            End Get
        End Property

        Public Property Province As String
            Get
                Return provinceField
            End Get

            Set(ByVal value As String)
                SetPropertyValue("Province", provinceField, value)
            End Set
        End Property

        Public Property ZipCode As String
            Get
                Return zipCodeField
            End Get

            Set(ByVal value As String)
                SetPropertyValue("ZipCode", zipCodeField, value)
            End Set
        End Property

        Public Property Country As String
            Get
                Return countryField
            End Get

            Set(ByVal value As String)
                SetPropertyValue("Country", countryField, value)
            End Set
        End Property

        Public Property City As String
            Get
                Return cityField
            End Get

            Set(ByVal value As String)
                SetPropertyValue("City", cityField, value)
            End Set
        End Property

        Public Property Address1 As String
            Get
                Return address1Field
            End Get

            Set(ByVal value As String)
                SetPropertyValue("Address1", address1Field, value)
            End Set
        End Property

        Public Property Address2 As String
            Get
                Return address2Field
            End Get

            Set(ByVal value As String)
                SetPropertyValue("Address2", address2Field, value)
            End Set
        End Property

        <Association>
        Public ReadOnly Property Persons As XPCollection(Of Contact)
            Get
                Return GetCollection(Of Contact)("Persons")
            End Get
        End Property

        <PersistentAlias("concat(Country, Province, City, ZipCode)")>
        Public ReadOnly Property FullAddress As String
            Get
                Return ObjectFormatter.Format("{Country}; {Province}; {City}; {ZipCode}", Me, EmptyEntriesMode.RemoveDelimiterWhenEntryIsEmpty)
            End Get
        End Property

        Public Sub New(ByVal session As Session)
            MyBase.New(session)
        End Sub

        Protected Overrides Function GetSequenceName() As String
            Return String.Concat(ClassInfo.FullName, "-", Province.Replace(" ", "_"))
        End Function
    End Class
End Namespace
